File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.py
from collections import Counter

def solution(participant, completion):
    answer = Counter(participant) - Counter(completion)
    return list(answer.keys())[0]

# 시간 복잡도 : O(N)
# Counter(participant) : 이름별 사람 수 카운팅
# Counter(completion) : 완주한 사람 수 카운팅

########### Counter가 머임? ###########
# dictionary처럼 동작하는 파이썬 내장 클래스
# list 같이 iterable 한 데이터 넣으면 자동으로 개수 세줌

# ex. Counter(['a', 'b', 'b', 'c']) >> Counter({'b': 2, 'a': 1, 'c': 1})

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java

import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";
        
        HashMap<String, Integer> map = new HashMap<>();
        for (String name : participant) {
	          map.put(name, map.getOrDefault(name, 0)+1);
				}
				for (String name : completion) {
	          map.put(name, map.get(name)-1);
				}
				
				for (String key : map.keySet()) {
					  if (map.get(key) > 0) {
							  answer = key;
							  break;
						}
				}
        return answer;
    }
}

/* HashMap.getOrDefault(Object key, Object defaultValue)*/
// 1) Java 8에 추가된 Collection API 메서드
// 2) 찾는 key가 존재하면 key의 value를 반환하고, 없을 경우(null) defaultValue를 반환